extensions/binding/org.eclipse.smarthome.binding.openweathermap/src/main/java/org/eclipse/smarthome/binding/openweathermap/internal/handler/OpenWeatherMapWeatherAndForecastHandler.java

```diff
@@ -0,0 +1,422 @@
+import org.eclipse.smarthome.core.thing.ThingStatusDetail;
+import org.eclipse.smarthome.core.thing.binding.ThingHandlerCallback;
+import org.eclipse.smarthome.core.thing.binding.builder.ChannelBuilder;
+import org.eclipse.smarthome.core.thing.type.ChannelGroupTypeUID;
+import org.eclipse.smarthome.core.types.State;
+import org.eclipse.smarthome.core.types.UnDefType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.JsonSyntaxException;
+
+/**
+ * The {@link OpenWeatherMapWeatherAndForecastHandler} is responsible for handling commands, which are sent to one of
+ * the channels.
+ *
+ * @author Christoph Weitkamp - Initial contribution
+ */
+@NonNullByDefault
+public class OpenWeatherMapWeatherAndForecastHandler extends AbstractOpenWeatherMapHandler {
+
+    private final Logger logger = LoggerFactory.getLogger(OpenWeatherMapWeatherAndForecastHandler.class);
+
+    private static final String CHANNEL_GROUP_HOURLY_FORECAST_PREFIX = "forecastHours";
+    private static final String CHANNEL_GROUP_DAILY_FORECAST_PREFIX = "forecastDay";
+    private static final Pattern CHANNEL_GROUP_HOURLY_FORECAST_PREFIX_PATTERN = Pattern
+            .compile(CHANNEL_GROUP_HOURLY_FORECAST_PREFIX + "([0-9]*)");
+    private static final Pattern CHANNEL_GROUP_DAILY_FORECAST_PREFIX_PATTERN = Pattern
+            .compile(CHANNEL_GROUP_DAILY_FORECAST_PREFIX + "([0-9]*)");
+
+    // keeps track of the parsed counts
+    private int forecastHours;
+    private int forecastDays;
+
+    private @Nullable OpenWeatherMapJsonWeatherData weatherData;
+    private @Nullable OpenWeatherMapJsonHourlyForecastData hourlyForecastData;
+    private @Nullable OpenWeatherMapJsonDailyForecastData dailyForecastData;
+
+    public OpenWeatherMapWeatherAndForecastHandler(Thing thing) {
+        super(thing);
+    }
+
+    @Override
+    public void initialize() {
+        super.initialize();
+        logger.debug("Initialize OpenWeatherMapWeatherAndForecastHandler handler {}.", getThing().getUID());
+        OpenWeatherMapWeatherAndForecastConfiguration config = getConfigAs(
+                OpenWeatherMapWeatherAndForecastConfiguration.class);
+
+        boolean configValid = true;
+        forecastHours = config.getForecastHours();
+        if (forecastHours < 0 || forecastHours > 120 || forecastHours % 3 != 0) {
+            forecastHours = 0;
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                    "@text/offline.conf-error-not-supported-number-of-hours");
+            configValid = false;
+        }
+        forecastDays = config.getForecastDays();
+        if (forecastDays < 0 || forecastDays > 16) {
+            forecastDays = 0;
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                    "@text/offline.conf-error-not-supported-number-of-days");
+            configValid = false;
+        }
+
+        if (configValid) {
```

Comment:
In initialization you dynamically create the channel groups for hourly and daily forecasts as defined in the config. Why are hourly forecast channels 03 - 24 predefined in the thing-type xml? Same for daily forecast channels 2 - 6?

Reply:
My initial idea was to have those channel groups by default and only change them dynamically if the user changes the thing configuration. Thus I had implemented a logic inside the `handleConfigurationUpdate()` method. But that did not work because when I tried to configure the things via textual configuration with different number of forecasted hours / days the method never had been triggered - which leads to a wrong count of channel groups. Afterwards I moved the logic to the initialization of the thing.

To be honest I do not think that it is the best way how to solve it. But for now I have no better solution in my mind. Maybe you have an idea how to do it?

So in general we now can remove them from the thing type definition.

Reply:
Like every time - once you start talking about it the idea starts to develop inside your mind.

I reworked the dynamically handling. Now only new channels will be added or superflous will be removed. This way the predefined ones in the thing-type xml as default makes sense again.